onnxruntime/core/optimizer/embed_layer_norm_fusion.cc

```diff
@@ -0,0 +1,183 @@
+    //Node& expand_node = *graph.GetNode(edges[1]->GetNode().Index());
+    //Node& shape_node = *graph.GetNode(edges[2]->GetNode().Index());
+    if (position_gather_node.GetOutputEdgesCount() != 1) {
+      continue;
+    }
+
+    // Get input "input_ids" from node.
+    NodeArg* input_ids = nullptr;
+    if (!graph_utils::NodeArgIsConstant(graph, *(word_gather_node.MutableInputDefs()[1])) &&
+        !graph_utils::IsGraphInput(graph, word_gather_node.MutableInputDefs()[1])) {
+      continue;
+    }
+    input_ids = word_gather_node.MutableInputDefs()[1];
+
+    // Get input "segment_ids" from node.
+    NodeArg* segment_ids = nullptr;
+    if (!graph_utils::NodeArgIsConstant(graph, *(segment_gather_node.MutableInputDefs()[1])) &&
+        !graph_utils::IsGraphInput(graph, segment_gather_node.MutableInputDefs()[1])) {
+      continue;
+    }
+    segment_ids = segment_gather_node.MutableInputDefs()[1];
+
+    // Get input "mask" from "ReduceSum" node.
+    NodeArg* mask = nullptr;
+    if (!graph_utils::NodeArgIsConstant(graph, *(reduce_sum_node.MutableInputDefs()[0])) &&
+        !graph_utils::IsGraphInput(graph, reduce_sum_node.MutableInputDefs()[0])) {
+      continue;
+    }
+    mask = reduce_sum_node.MutableInputDefs()[0];
+
+    const std::vector<NodeArg*> embed_layer_norm_input_defs{
+        input_ids,
+        segment_ids,
+        mask,
+        word_gather_node.MutableInputDefs()[0],
+        position_gather_node.MutableInputDefs()[0],
+        segment_gather_node.MutableInputDefs()[0],
+        skip_ln_node.MutableInputDefs()[2],
+        skip_ln_node.MutableInputDefs()[3]};
+    Node& embed_layer_norm_node = graph.AddNode(graph.GenerateNodeName("EmbedLayerNormalization"),
+                                                "EmbedLayerNormalization", 
+                                                "fused EmbedLayerNorm subgraphs ", 
+                                                embed_layer_norm_input_defs,
+                                                {skip_ln_node.MutableOutputDefs()[0], reduce_sum_node.MutableOutputDefs()[0]},
+                                                {}, kMSDomain);
+
+    // Assign provider to this new node. Provider should be same as the provider for old node.
+    embed_layer_norm_node.SetExecutionProviderType(skip_ln_node.GetExecutionProviderType());
+
+    // move input edges to add (first in list) across to the embed_layer_norm_node.
+    // move output definitions and output edges to embed_layer_norm_node.
+    // remove all the other nodes.
+    std::vector<Node*> nodes_to_remove{
+        graph.GetNode(word_gather_node.Index()),
+        graph.GetNode(position_gather_node.Index()),
+        //graph.GetNode(shape_node.Index()),
+        //graph.GetNode(expand_node.Index()),
+        graph.GetNode(segment_gather_node.Index()),
+        graph.GetNode(add_node.Index()),
+        graph.GetNode(reduce_sum_node.Index()),
```

Comment:
reduce_sum could be removed only if its output edge count is same as attention node count.